mdjs-layout/src/MdjsLayout.js

```diff
@@ -0,0 +1,421 @@
+import { css, html, LitElement } from 'lit';
+import { unsafeHTML } from 'lit/directives/unsafe-html.js';
+import { ThemeMixin } from 'dark-theme-utils';
+
+export class MdjsLayout extends ThemeMixin(LitElement) {
+  static get properties() {
+    return {
+      content: { attribute: false },
+    };
+  }
+
+  static get styles() {
+    return css`
+      :host {
+        display: block;
+        font-family: 'Inter var', ui-sans-serif, system-ui, -apple-system,
+          BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial,
+          'Noto Sans', sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji',
+          'Segoe UI Symbol', 'Noto Color Emoji';
+        max-width: 1200px;
+        margin: 50px auto;
+        padding: 0 30px;
+        color: #171717;
+        background-color: #FAFAFA;
+      }
+
+      :host([theme='dark']) {
+        color: #FAFAFA;
+        background-color: #262626;
+      }
+
+      h1,
+      h2,
+      h3,
+      h4,
+      h5,
+      h6 {
+        color: #171717;
+      }
+
+      :host([theme='dark']) h1,
+      :host([theme='dark']) h2,
+      :host([theme='dark']) h3,
+      :host([theme='dark']) h4,
+      :host([theme='dark']) h5,
+      :host([theme='dark']) h6 {
+        color: var(--heading-color-dark);
+      }

```

Comment:
IIUC, you want to make light/dark switcher generic for all MDJS users, which I fully support

but we need to find a good way to integrate with all popular WC frameworks, and right now it's a challenge, because in FAST it's different, it's global, so there is no need to add a mixin to every component

look at this code which I use in my documentation

```js preview-story
import { baseLayerLuminance } from '@divriots/starter-furious';
import { StandardLuminance } from '@microsoft/fast-components';

class ModeSwitcher extends FASTElement {
  static definition = {
    name: 'mode-switcher',
    template: html`<button @click=${(x) => x.switch()}>Click to switch the mode</button>`,
  };

  switch() {
    const currentLuminanceValue = baseLayerLuminance.getValueFor(document.body);
    const newLuminanceValue =
      currentLuminanceValue === StandardLuminance.LightMode ? StandardLuminance.DarkMode : StandardLuminance.LightMode;
    baseLayerLuminance.setValueFor(document.body, newLuminanceValue);
  }
}

FASTElement.define(ModeSwitcher);

export const modeSwitcherStory = () => html`<mode-switcher></mode-switcher>`;
```

under the hood it translates to CSS Custom Props which are defined on some element (`document.body` in this case), and they just cascade normally and components natively update, because CSS Custom Props have updated

Reply:
actually I now realise that it's not too important how FAST implements light/dark modes and more important how it expects it to be switched

I see the only way to generalize it: provide a hook method for each DS to switch the mode

Reply:
Yeah I looked into it and correct me if I'm wrong, but it's super similar to how I do color theming in simba (this is separate from dark-mode btw):

```js
// Global color theme setter
export const setTheme = (_theme) => {
  theme = _theme;
  localStorage.setItem('simba-color', theme);
  const obj = {
    '--color-primary-50': colors[theme][50],
    '--color-primary-100': colors[theme][100],
    '--color-primary-200': colors[theme][200],
    '--color-primary-300': colors[theme][300],
    '--color-primary-400': colors[theme][400],
    '--color-primary-500': colors[theme][500],
    '--color-primary-600': colors[theme][600],
    '--color-primary-700': colors[theme][700],
    '--color-primary-800': colors[theme][800],
    '--color-primary-900': colors[theme][900],
  };

  for (let [key, value] of Object.entries(obj)) {
    document.documentElement.style.setProperty(key, value);
  }
};
```

```js
import { setTheme } from '~/themes';

setTheme('amber');
```

Then for a component to be "sensitive" to this color theme, you just make sure it uses one or multiple of these color-primary CSS custom props. 

You could certainly do the same thing for dark-mode.

My arguments for why this approach is wrong for dark mode is as follows. 

When my user imports my button component, I want to make sure that this button component already has dark-mode support fully out of the box, and my user doesn't need to set any CSS custom props for it. This makes the component encapsulated and it makes it possible for me to make the claim it has dark-mode support without extra effort needed. 

The disadvantage is that I assumed a convention, a "theme" attribute on the `html` element that is either "light" or "dark".

But basically the DX is with custom props:

```js
import 'simba-button'; 
import { setDarkMode } from 'dark-mode-global';

setDarkMode(true); // changes value of custom CSS prop under the hood

export default html`<simba-button></simba-button>`;
```

vs when the button encapsulates its darkmode support fully:

```js
import 'simba-button'; 

// agnostic to how the dark mode is set, but responds to it anyway

export default html`<simba-button></simba-button>`;
```

With regards to making a hook, what about something like this:

`mdjs.config.js`:
```js
import { baseLayerLuminance } from '@divriots/starter-furious';
 
export default {
  layout: (content, context) => html`<layout>...</layout>`,
  onDarkModeChanged: (mode) => { // 'dark' | 'light'
    const currentLuminanceValue = baseLayerLuminance.getValueFor(document.body);
    const newLuminanceValue =
    currentLuminanceValue === StandardLuminance.LightMode ? StandardLuminance.DarkMode : StandardLuminance.LightMode;
    baseLayerLuminance.setValueFor(document.body, newLuminanceValue);
    
  }
};
```

Meaning the hook is called when we change the theme in our backlight editor:
![image](https://user-images.githubusercontent.com/36734656/132193482-5c522669-6587-49f9-b2a4-b3b7c4fef703.png)

That would require some changes, by adding a new button, because right now that theme button only changes the background-color of the `<body>` element. Since I already created a theme-toggler CE in `dark-theme-utils`, I would like to propose extending that CE to our own theme-toggler and add that to the UI at the bottom, and making it so that the onDarkModeChanged hook is called so that design systems can configure how to respond to it, if they do not use the convention that I am assuming with a "theme" attribute on the `html` element. 

Thoughts?

Reply:
Each approach has pros and cons. Let's focus on making this work for FAST too, and not focus on how theming is implemented under the hood.

```js
  onDarkModeChanged: (mode) => { // 'dark' | 'light'
    const currentLuminanceValue = baseLayerLuminance.getValueFor(document.body);
    const newLuminanceValue =
    currentLuminanceValue === StandardLuminance.LightMode ? StandardLuminance.DarkMode : StandardLuminance.LightMode;
    baseLayerLuminance.setValueFor(document.body, newLuminanceValue);
  }
```

Given Backlight will do the actual DOM update (e.g. html attribute change), it's logical to me that the config will support such a hook. I'm trying to think of the best name, right now I'm not confident that we should call this DarkMode, because we have 2: light and dark, so maybe call at just a mode making `onThemeModeChanged` or smth. Thoughts?

In addition to that I'd like to control how it's initialised or how page and the DS mode is initialised. We ether need to set the default ion the config (probably worse), or call this hook immediately on page load (probably better) even if user didn't click on it yet.

> That would require some changes, by adding a new button, because right now that theme button only changes the background-color of the <body> element. 

looks logical to me, I don't think the current one is designed to do what we need, but I'd check with people who made it


Last, but not the least and kinda unrelated. The more we discuss `mdjs.config.js`, the more I dislike the name. I've said that earlier already that we need to contribute all this ideas back to it or rename it and make it `studio.config.js` or whatever else Backlight/divriots specific. cc @muryoh @georges-gomes @gqio

Reply:
![image](https://user-images.githubusercontent.com/5250572/132233533-c7f838b1-e8e5-4740-b370-f3e63c547906.png)

The background button is legacy and we should remove it as some point to offer "backgrounds" feature from storybook. https://storybook.js.org/docs/react/essentials/backgrounds

But backgrounds in storybook is just a way to control the background and not a dark/light switch.

I think that the right way to do dark/light switch *for the moment* is to implement a decorator to use on every stories that propose a panel of options and one of them is dark/light switch.
Check what Adobe has done on WCD https://webcomponents.dev/edit/collection/fO75441E1Q5ZlI0e9pgq/Zjc3o94DWuBkT4ve3dny/src/index.stories.js?branch=master



Later on, we need to support Storybook addon hooks to add toolbar buttons. And then stories can have this button on the toolbar like https://storybook.js.org/addons/storybook-darkmode-vue/
(Backgrounds in storybook is an addon https://storybook.js.org/addons/@storybook/addon-backgrounds/)
And I think we can create a FAST darkmode toggle or a class darkmode toggle (for SIMBA and others) and offer them in an open source story package "a la" dockits

Reply:
@georges-gomes  

- Are Storybook decorators already working with MDJS as well? same syntax? I'm not entirely familiar with the MDJS/Storybook cross-integration in backlight, if there is any.
- Perhaps we can put this temp decorator (until we have addon) for themes in `.storybook/preview.js` instead of asking from the user they include it in all their story files separately, which is what I see in Adobe SWC. https://storybook.js.org/docs/react/writing-stories/decorators#context-for-mocking seems possible from the docs